protocol/daemons/server/types/health_monitor.go

```diff
@@ -0,0 +1,293 @@
+	// been unhealthy. If this timestamp is nil, then the error streak ended before it could trigger a callback.
+	// Access to firstFailureInStreak is synchronized.
+	firstFailureInStreak timestampWithError
+
+	// timeProvider is used to get the current time. It is added as a field so that it can be mocked in tests.
+	// Access to timeProvider is synchronized.
+	timeProvider libtime.TimeProvider
+}
+
+// Poll executes a health check for the health checkable service. If the service has been unhealthy for longer than the
+// maximum acceptable unhealthy duration, the callback function is executed.
+// This method is publicly exposed for testing. This method is synchronized.
+func (hc *healthChecker) Poll() {
+	hc.lock.Lock()
+	defer hc.lock.Unlock()
+
+	// Don't return an error if the monitor has been disabled.
+	err := hc.healthCheckable.HealthCheck()
+	now := hc.timeProvider.Now()
+	if err == nil {
+		hc.mostRecentSuccess = now
+		// Whenever the service is healthy, reset the first failure in streak timestamp.
+		hc.firstFailureInStreak.Reset()
+	} else if hc.firstFailureInStreak.IsZero() {
+		// Capture the timestamp of the first failure in a new streak.
+		hc.firstFailureInStreak.Update(now, err)
+	}
+
+	// If the service has been unhealthy for longer than the maximum acceptable unhealthy duration, execute the
+	// callback function.
+	if !hc.firstFailureInStreak.IsZero() &&
+		now.Sub(hc.firstFailureInStreak.Timestamp()) >= hc.maximumAcceptableUnhealthyDuration {
+		hc.unhealthyCallback(hc.firstFailureInStreak.Error())
+	}
+	// Schedule the next poll.
+	hc.timer.Reset(hc.pollFrequency)
+}
+
+// Stop stops the health checker. This method is synchronized.
+func (hc *healthChecker) Stop() {
+	hc.lock.Lock()
+	defer hc.lock.Unlock()
+
+	hc.timer.Stop()
+}
+
+// StartNewHealthChecker creates and starts a new health checker for a health checkable service.
+func StartNewHealthChecker(
+	healthCheckable types.HealthCheckable,
+	pollFrequency time.Duration,
+	unhealthyCallback func(error),
+	timeProvider libtime.TimeProvider,
+	maximumAcceptableUnhealthyDuration time.Duration,
+	startupGracePeriod time.Duration,
+) *healthChecker {
+	checker := &healthChecker{
+		healthCheckable:                    healthCheckable,
+		pollFrequency:                      pollFrequency,
+		unhealthyCallback:                  unhealthyCallback,
+		timeProvider:                       timeProvider,
+		maximumAcceptableUnhealthyDuration: maximumAcceptableUnhealthyDuration,
+	}
+	// The first poll is scheduled after the startup grace period to allow the service to initialize.
+	checker.timer = time.AfterFunc(startupGracePeriod, checker.Poll)
+
+	return checker
+}
+
+// HealthMonitor monitors the health of daemon services, which implement the HealthCheckable interface. If a
+// registered health-checkable service sustains an unhealthy state for the maximum acceptable unhealthy duration,
+// the monitor will execute a callback function.
+type HealthMonitor struct {
```

Comment:
nit: we could break this file into `health_checker.go` and `health_monitor.go` but will leave it up to you!